README.md
Neutrino
========
Neutrino is a flexible jQuery based slideshow plugin that requires minimum markup. Styles are all in the LESS/CSS files and so both the styles and the JavaScript are required for Neutrino to work.

Dependencies
------------
- [jQuery](http://jquery.com) {Use jQuery < 2.0 if you want to support IE8}
- [TweenMax](http://www.greensock.com/gsap-js/)


Basic usage
-----------

First, be sure to include all the above libraries. Then, include Neutrino :
````html
<script src="js/neutrino.jquery.js" type="text/javascript"></script>
````

To set up Neutrino, simply add a div tag with the class "neutrino" and then add as many slides as you want by
adding div tags with the class "slide". You need to have at least to slides for Neutrino to work because it 
would be quite silly to make a slideshow with only one slide, wouldn't it?
````html
<div class="neutrino">
	<div class="slide"></div>
	<div class="slide"></div>
</div>
````

Using jQuery, setup Neutrino on the selected element :
````js
$(".neutrino").neutrino();
````

You can customize Neutrino using an options argument (more details on all options will follow):
````js
$(".neutrino").neutrino(options);
````

Options
-------
````js
options: {
	transitionType: 'slide', // A string representing the type of transition, currently, only 'slide' is supported.
	transitionTime: 0.75, // The time in SECONDS that the animation between each slides will take.
	timer: 3500, // The time in MILLISECONDS between each animations. 3500 is default. If set to 0, there will not be a timed loop.
	hasArrows: false, // False is default. If set to true, Neutrino will add arrows on each sides of the slideshow.
	hasNav: false // False is default. If set to true, Neutrino will add a navigation at the bottom of the slideshow.
}
````


CHANGELOG.md
Patch notes
-----------------------------

#####-  0.6.0 : 
- Major update of the script to make it a better jQuery plugin.
- Update of the folder structures.

#####-  0.5.0 : 
- Deleted AMD folder. No longer needed.
- Major update to the README.md file and added a CHANGELOG.md file.
- Changed the remote repo. It is now a LaGrangeMtl repo.
- neutrino.js has been replaced with neutrino.jquery.js has it is now a jQuery plugin.
- Skipped a few logs. 
-- Major changes in the code, now having a few closures to help prevent bugs.

#####-  0.3.2 : 
- Arrows are now added by the JS, like the navigation. Default is set to false.

#####-  0.3.1 : 
- Updated README.md and reorganised the folders so that the demo is separated from the Neutrino folders

#####-  0.3.0 : 
- Navigation is working.

#####-  0.2.0 : 
- Arrows on each side are working.

#####-  0.1.0 : 
- First working version. Only working with the slide animation right now.

#####-  0.0.2 : 
- Updated README.md

#####-  0.0.1 : 
- Started the project, and it feels awesome.

scripts/neutrino.js
/*The MIT License (MIT)

Copyright (c) 2013 Nicolas Poirier-Barolet

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

function Neutrino() {
	var neutrino = {
		params : {},

		//=====================================================================
		// init : Public Function
		//
		// @params : settings
		//		Custom settings of the user. May be an empty object.
		//
		// @params : context
		//		If defined, will be the root. This is to be able to have multiple
		//		neutrino slideshow in your page. Though I do not recommend it, it
		//		is possible.
		//
		// This function initialize the slideshow process. Variables are set up 
		// here and the Parameters of the whole slideshow also. After the 
		// setting is done, it starts the timer.
		//=====================================================================
		init : function(settings, context) {
			this.root = context || $('.neutrino');
			this.slides = this.root.find('.slide');
			this.arrows = this.root.find('.arrow');
			this.nav = this.root.find('nav');
			this.navButtons = undefined;
			this.timer = undefined;

			this._setParams(settings);
			this.currentIndex = 0;
			this.direction = 1;

			if(this.params.hasArrows) {
				this._createArrows();
			}

			if(this.params.hasNav) {
				this._createNav();
				this._updateNav();
			}

			// Check with settings.timer and not with this.params.timer
			// because that if you put 0, it is the same as false, and
			// therefore, this.params.timer will equal the default value
			if(settings.timer > 0) {
				this._initSlides();
				this._setTimer();
			}
			else {
				this.params.timer = false;
				this._initSlides();
			}
		},

		//=====================================================================
		// _setParams : Private Function
		//
		// @params : settings
		//		Custom settings of the user. May be an empty object.
		//
		// If settings is defined, it's properties will be used and if not, the
		// default settings will be used instead. Settings might be omitted if
		// you want to use some defaults.
		//=====================================================================
		_setParams : function(settings) {
			this.params = {
				transitionType: settings.transitionType || 'slide',
				transitionTime: settings.transitionTime || '0.75',
				slideWidth: this.slides.eq(0).width(),
				timer: settings.timer || '3500',
				hasArrows: settings.hasArrows || false,
				hasNav: settings.hasNav || false
			};
		},

		//=====================================================================
		// _setArrowEvents : Private Function
		//
		// Will set the click events on the arrows. Direction is set by the
		// data attribute on the tags.
		//=====================================================================
		_setArrowEvents : function() {
			this.arrows.on('click.neutrino', function(e){
				clearTimeout(this.timer);

				this.direction = $(e.target).data('direction');

				this.arrows.off('.neutrino');
				this._initSlides(e);
				this._changeSlide();
			}.bind(this))
		},

		//=====================================================================
		// _createArrows : Private Function
		//
		// Creates the arrows for the slideshow. Uses <div> tags, 
		// and sets the direction by using the data attribute.
		//=====================================================================
		_createArrows : function(){
			var arrowsMarkup = '<div class="arrow left" data-direction="-1"></div>';
			arrowsMarkup += '<div class="arrow right" data-direction="1"></div>';

			this.root.append(arrowsMarkup);
			this.arrows = this.root.find('.arrow');

			this._setArrowEvents();
		},

		//=====================================================================
		// _setNavEvents : Private Function
		//
		// Will set the click events on the nav. If you click on an already
		// active slide, the function will return, thus making the whole
		// process wait for another click or the timer to change the slide.
		//=====================================================================
		_setNavEvents : function() {
			var _self = this;
			this.navButtons.on('click.neutrino', function(e){
				// If is already active, wait for the timer to change slide
				if($(e.target).hasClass('active')){
					return;
				}
				else {
					clearTimeout(_self.timer);

					_self.direction = 0;

					_self._initSlides(e);
					_self._changeSlide();
				}
			});
		},

		//=====================================================================
		// _updateNav : Private Function
		//
		// Updates the navButtons style.
		//=====================================================================
		_updateNav : function() {
			this.navButtons.removeClass('active');
			this.navButtons.eq(this.currentIndex).addClass('active');
		},

		//=====================================================================
		// _createNav : Private Function
		//
		// Creates the navigation for the slideshow. Uses <li> tags, 
		// and sets the width of the <ul> for it to be centered.
		//=====================================================================
		_createNav : function(){
			var nbOfSlides = this.slides.length;
			var nav = '<nav><ul>';

			for(var i=0; i < this.slides.length; i++) {
				nav += '<li></li>';
			}

			nav += '</ul></nav>';

			this.root.append(nav);
			this.nav = this.root.find('nav');
			this.navButtons = this.nav.find('li');

			var liWidth = this.navButtons.eq(0).width();
			var liMargin = this.navButtons.eq(1).css('margin-left');
			liMargin = liMargin.substring(0, liMargin.length - 2);

			this.nav.find('ul').css({width: (liWidth * nbOfSlides) + (liMargin * (nbOfSlides - 1)) + "px"})

			this._setNavEvents();
		},

		//=====================================================================
		// _initSlides : Private Function
		//
		// @params : e
		//		If not defined, it means no click event was done to get here
		//		therefore, the direction should be 1, which is equal to right
		//		to left. Please note that not all of the animations will make
		//		use of the direction parameter. i.e. Fade in/out
		//
		//		If e is defined, a click was made. If it was on the nav, the
		//		direction property will be of 0. The nextIndex will then become
		//		the targeted nav button. Otherwise, the direction will be equal
		//		to what the arrow event set it to [see _setArrowEvents()]
		//=====================================================================
		_initSlides : function(e){
			this.currentSlide = this.slides.eq(this.currentIndex);
			this.nextIndex = this.currentIndex + this.direction;

			if(e){
				// If this direction == 0, it means that we clicked on the nav
				// buttons
				if(this.direction == 0) {
					this.nextIndex = this.navButtons.index($(e.target));

					if(this.nextIndex < this.currentIndex)
						this.direction = -1;
					else
						this.direction = 1;
				}
			}
			else {
				this.direction = 1;
			}

			if(this.nextIndex >= this.slides.length)
				this.nextIndex = 0;
			else if (this.nextIndex < 0)
				this.nextIndex = this.slides.length - 1;

			this.nextSlide = this.slides.eq(this.nextIndex);

			this.slides.hide();
			this.currentSlide.show();
		},

		//=====================================================================
		// _changeSlide : Private Function
		//
		// This is where the slides are changed. For the whole process, the 
		// arrows and the nav buttons will be disabled. They will be enabled
		// again after the animations are done. Depending on the transitionType
		// property of the slideshow, it will call the right animation function
		//=====================================================================
		_changeSlide : function(){
			var animation;

			if(this.params.hasNav)
				this.navButtons.off('.neutrino');

			if(this.params.transitionType == 'slide')
				animation = this._slide();
			else
				alert('Only the SLIDE effect is working right now.');

			var _self = this;
			animation.done(function(){
				if(_self.params.hasNav){
					_self._updateNav();
					_self._setNavEvents();
				}

				if(_self.params.hasArrows)
					_self._setArrowEvents();
				
				if(_self.params.timer) {
					_self.direction = 1;
					_self._setTimer();
				}
			})
		},

		//=====================================================================
		// _slide : Private Function
		//
		// @returns : $.Deferred();
		//
		// Function to be called when the transitionType property is set to
		// 'slide'.
		//=====================================================================
		_slide : function(){
			var firstSlide = $.Deferred();
			var secondSlide = $.Deferred();
			var animationDeferred = $.Deferred();

			this.nextSlide
				.css({left: (this.params.slideWidth * this.direction)})
				.show();

			var _self = this;
			TweenMax.to(this.currentSlide, this.params.transitionTime, {
				left: "+="+(this.params.slideWidth * (this.direction * -1)),
				
				onComplete:function(){
					firstSlide.resolve();

					_self.currentSlide
						.hide()
						.css({left:0});
				}
			});

			TweenMax.to(this.nextSlide, this.params.transitionTime, {
				left: 0,
				
				onComplete:function(){
					secondSlide.resolve();
				}
			});

			_self = this;
			$.when(firstSlide, secondSlide).then(function(){
				_self.currentIndex = _self.nextIndex;

				animationDeferred.resolve();
			});

			return animationDeferred;
		},

		//=====================================================================
		// _setTimer : Private Function
		//
		// Sets the timeout function to call _initSlides after each tick.
		//=====================================================================
		_setTimer : function(){
			var _self = this;

			this.timer = setTimeout(function(){
				_self._initSlides();
				_self._changeSlide();
			}, this.params.timer);
		}
	}

	return neutrino;
};

source/js/neutrino.jquery.js
/*The MIT License (MIT)

Copyright (c) 2013 Nicolas Poirier-Barolet

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

;(function($) {
	var Neutrino = {
		//=====================================================================
		// init : Public Function
		//
		// @params : options
		//		Custom settings of the user. May be an empty object.
		//
		// @params : context
		//		Element that will be defined as the root of the slideshow
		//
		// This function initialize the slideshow process. Variables are set up 
		// here and the Parameters of the whole slideshow also. After the 
		// setting is done, it starts the timer.
		//=====================================================================
		init : function(options, context) {
			this.root = $(context);
			this.slides = this.root.find('.slide');

			if(this.slides.length <= 1){
				console.log("Neutrino : There is only " + this.slides.length + " slides set in the markup. Neutrino needs at least 2 slides to work.");
				return;
			}

			// Default values
			this.options = {
				transitionType: 'slide',
				transitionTime: 0.75,
				slideWidth: this.slides.eq(0).width(),
				timer: 3500,
				hasArrows: false,
				hasNav: false
			};

			this.options = $.extend({},this.options,options);
			
			this._build();

			return this;
		},

		//=====================================================================
		// _build : Private Function
		//
		// Sets various properties to the Neutrino object that will be used
		// later. Creates Navigation and Arrows if needed and starts the slider
		// with a timer or not, depending of the options. 
		//=====================================================================
		_build : function() {
			this.arrows = this.root.find('.arrow');
			this.nav = this.root.find('nav');
			this.navButtons = undefined;
			this.timer = undefined;

			this.currentIndex = 0;
			this.direction = 1;

			if(this.options.hasArrows) {
				this._createArrows();
			}

			if(this.options.hasNav) {
				this._createNav();
				this._updateNav();
			}

			if(this.options.timer > 0 || this.options == undefined) {
				this._initSlides();
				this._setTimer();
			}
			else {
				this.options.timer = false;
				this._initSlides();
			}
		},

		//=====================================================================
		// _setArrowEvents : Private Function
		//
		// Will set the click events on the arrows. Direction is set by the
		// data attribute on the tags.
		//=====================================================================
		_setArrowEvents : function() {
			this.arrows.on('click.neutrino', function(e){
				clearTimeout(this.timer);

				this.direction = $(e.target).data('direction');

				this.arrows.off('.neutrino');
				this._initSlides(e);
				this._changeSlide();
			}.bind(this))
		},

		//=====================================================================
		// _createArrows : Private Function
		//
		// Creates the arrows for the slideshow. Uses <div> tags, 
		// and sets the direction by using the data attribute.
		//=====================================================================
		_createArrows : function(){
			var arrowsMarkup = '<div class="arrow left" data-direction="-1"></div>';
			arrowsMarkup += '<div class="arrow right" data-direction="1"></div>';

			this.root.append(arrowsMarkup);
			this.arrows = this.root.find('.arrow');

			this._setArrowEvents();
		},

		//=====================================================================
		// _setNavEvents : Private Function
		//
		// Will set the click events on the nav. If you click on an already
		// active slide, the function will return, thus making the whole
		// process wait for another click or the timer to change the slide.
		//=====================================================================
		_setNavEvents : function() {
			var _self = this;
			this.navButtons.on('click.neutrino', function(e){
				// If is already active, wait for the timer to change slide
				if($(e.target).hasClass('active')){
					return;
				}
				else {
					clearTimeout(_self.timer);

					_self.direction = 0;

					_self._initSlides(e);
					_self._changeSlide();
				}
			});
		},

		//=====================================================================
		// _updateNav : Private Function
		//
		// Updates the navButtons style.
		//=====================================================================
		_updateNav : function() {
			this.navButtons.removeClass('active');
			this.navButtons.eq(this.currentIndex).addClass('active');
		},

		//=====================================================================
		// _createNav : Private Function
		//
		// Creates the navigation for the slideshow. Uses <li> tags, 
		// and sets the width of the <ul> for it to be centered.
		//=====================================================================
		_createNav : function(){
			var nbOfSlides = this.slides.length;
			var nav = '<nav><ul>';

			for(var i=0; i < this.slides.length; i++) {
				nav += '<li></li>';
			}

			nav += '</ul></nav>';

			this.root.append(nav);
			this.nav = this.root.find('nav');
			this.navButtons = this.nav.find('li');

			var liWidth = this.navButtons.eq(0).width();
			var liMargin = this.navButtons.eq(1).css('margin-left');
			liMargin = liMargin.substring(0, liMargin.length - 2);

			this.nav.find('ul').css({width: (liWidth * nbOfSlides) + (liMargin * (nbOfSlides - 1)) + "px"})

			this._setNavEvents();
		},

		//=====================================================================
		// _initSlides : Private Function
		//
		// @params : e
		//		If not defined, it means no click event was done to get here
		//		therefore, the direction should be 1, which is equal to right
		//		to left. Please note that not all of the animations will make
		//		use of the direction parameter. i.e. Fade in/out
		//
		//		If e is defined, a click was made. If it was on the nav, the
		//		direction property will be of 0. The nextIndex will then become
		//		the targeted nav button. Otherwise, the direction will be equal
		//		to what the arrow event set it to [see _setArrowEvents()]
		//=====================================================================
		_initSlides : function(e){
			this.currentSlide = this.slides.eq(this.currentIndex);
			this.nextIndex = this.currentIndex + this.direction;

			if(e){
				// If this direction == 0, it means that we clicked on the nav
				// buttons
				if(this.direction == 0) {
					this.nextIndex = this.navButtons.index($(e.target));

					if(this.nextIndex < this.currentIndex)
						this.direction = -1;
					else
						this.direction = 1;
				}
			}
			else {
				this.direction = 1;
			}

			if(this.nextIndex >= this.slides.length)
				this.nextIndex = 0;
			else if (this.nextIndex < 0)
				this.nextIndex = this.slides.length - 1;

			this.nextSlide = this.slides.eq(this.nextIndex);

			this.slides.hide();
			this.currentSlide.show();
		},

		//=====================================================================
		// _changeSlide : Private Function
		//
		// This is where the slides are changed. For the whole process, the 
		// arrows and the nav buttons will be disabled. They will be enabled
		// again after the animations are done. Depending on the transitionType
		// property of the slideshow, it will call the right animation function
		//=====================================================================
		_changeSlide : function(){
			var animation;

			if(this.options.hasNav)
				this.navButtons.off('.neutrino');

			if(this.options.transitionType == 'slide')
				animation = this._slide();
			else
				alert('Only the SLIDE effect is working right now.');

			var _self = this;
			animation.done(function(){
				if(_self.options.hasNav){
					_self._updateNav();
					_self._setNavEvents();
				}

				if(_self.options.hasArrows)
					_self._setArrowEvents();
				
				if(_self.options.timer) {
					_self.direction = 1;
					_self._setTimer();
				}
			})
		},

		//=====================================================================
		// _slide : Private Function
		//
		// @returns : $.Deferred();
		//
		// Function to be called when the transitionType property is set to
		// 'slide'.
		//=====================================================================
		_slide : function(){
			var firstSlide = $.Deferred();
			var secondSlide = $.Deferred();
			var animationDeferred = $.Deferred();

			this.nextSlide
				.css({left: (this.options.slideWidth * this.direction)})
				.show();

			var _self = this;
			TweenMax.to(this.currentSlide, this.options.transitionTime, {
				left: "+="+(this.options.slideWidth * (this.direction * -1)),
				
				onComplete:function(){
					firstSlide.resolve();

					_self.currentSlide
						.hide()
						.css({left:0});
				}
			});

			TweenMax.to(this.nextSlide, this.options.transitionTime, {
				left: 0,
				
				onComplete:function(){
					secondSlide.resolve();
				}
			});

			_self = this;
			$.when(firstSlide, secondSlide).then(function(){
				_self.currentIndex = _self.nextIndex;

				animationDeferred.resolve();
			});

			return animationDeferred;
		},

		//=====================================================================
		// _setTimer : Private Function
		//
		// Sets the timeout function to call _initSlides after each tick.
		//=====================================================================
		_setTimer : function(){
			var _self = this;

			this.timer = setTimeout(function(){
				_self._initSlides();
				_self._changeSlide();
			}, this.options.timer);
		}
	};

	$.fn.neutrino = function(options) {
		if(this.length) {
			return this.each(function(){
				var neutrino = Object.create(Neutrino);
				neutrino.init(options, this);
				$.data(this, 'neutrino', neutrino);
			})
		}
	};
})(jQuery)

AMD/Neutrino.js
/*The MIT License (MIT)

Copyright (c) 2013 Nicolas Poirier-Barolet

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

define(
	[
		'jquery',
		'TweenMax'
	],
	function($) {
		var Neutrino = {
			params : {},

			//=====================================================================
			// init : Public Function
			//
			// @params : settings
			//		Custom settings of the user. May be an empty object.
			//
			// @params : context
			//		If defined, will be the root. This is to be able to have multiple
			//		neutrino slideshow in your page. Though I do not recommend it, it
			//		is possible.
			//
			// This function initialize the slideshow process. Variables are set up 
			// here and the Parameters of the whole slideshow also. After the 
			// setting is done, it starts the timer.
			//=====================================================================
			init : function(settings, context) {
				this.root = context || $('.neutrino');
				this.slides = this.root.find('.slide');
				this.arrows = this.root.find('.arrow');
				this.nav = this.root.find('nav');

				this._setParams(settings);
				this.currentIndex = 0;
				this.direction = 1;

				if(this.params.hasArrows) {
					this._createArrows();
				}

				if(this.params.hasNav) {
					this._createNav();
					this._updateNav();
				}

				// Check with settings.timer and not with this.params.timer
				// because that if you put 0, it is the same as false, and
				// therefore, this.params.timer will equal the default value
				if(settings.timer > 0) {
					this._setTimer();
				}
				else {
					this.params.timer = false;
					this._initSlides();
				}
			},

			//=====================================================================
			// _setParams : Private Function
			//
			// @params : settings
			//		Custom settings of the user. May be an empty object.
			//
			// If settings is defined, it's properties will be used and if not, the
			// default settings will be used instead. Settings might be omitted if
			// you want to use some defaults.
			//=====================================================================
			_setParams : function(settings) {
				this.params = {
					transitionType: settings.transitionType || 'slide',
					transitionTime: settings.transitionTime || '0.75',
					slideWidth: this.slides.eq(0).width(),
					timer: settings.timer || '3500',
					hasArrows: settings.hasArrows || false,
					hasNav: settings.hasNav || false
				};
			},

			//=====================================================================
			// _setArrowEvents : Private Function
			//
			// Will set the click events on the arrows. Direction is set by the
			// data attribute on the tags.
			//=====================================================================
			_setArrowEvents : function() {
				this.arrows.on('click.neutrino', function(e){
					clearTimeout(this.timer);

					this.direction = $(e.target).data('direction');

					this.arrows.off('.neutrino');
					this._initSlides(e);
					this._changeSlide();
				}.bind(this))
			},

			//=====================================================================
			// _createArrows : Private Function
			//
			// Creates the arrows for the slideshow. Uses <div> tags, 
			// and sets the direction by using the data attribute.
			//=====================================================================
			_createArrows : function(){
				var arrowsMarkup = '<div class="arrow left" data-direction="-1"></div>';
				arrowsMarkup += '<div class="arrow right" data-direction="1"></div>';

				this.root.append(arrowsMarkup);
				this.arrows = this.root.find('.arrow');

				this._setArrowEvents();
			},

			//=====================================================================
			// _setNavEvents : Private Function
			//
			// Will set the click events on the nav. If you click on an already
			// active slide, the function will return, thus making the whole
			// process wait for another click or the timer to change the slide.
			//=====================================================================
			_setNavEvents : function() {
				this.navButtons.on('click.neutrino', function(e){
					// If is already active, wait for the timer to change slide
					if($(e.target).hasClass('active')){
						return;
					}
					else {
						clearTimeout(this.timer);

						this.direction = 0;

						this.navButtons.off('.neutrino');
						this._initSlides(e);
						this._changeSlide();
					}
				}.bind(this))
			},

			//=====================================================================
			// _updateNav : Private Function
			//
			// Updates the navButtons style.
			//=====================================================================
			_updateNav : function() {
				this.navButtons.removeClass('active');
				this.navButtons.eq(this.currentIndex).addClass('active');
			},

			//=====================================================================
			// _createNav : Private Function
			//
			// Creates the navigation for the slideshow. Uses <li> tags, 
			// and sets the width of the <ul> for it to be centered.
			//=====================================================================
			_createNav : function(){
				var nbOfSlides = this.slides.length;
				var nav = '<nav><ul>';

				for(var i=0; i < this.slides.length; i++) {
					nav += '<li></li>';
				}

				nav += '</ul></nav>';

				this.root.append(nav);
				this.nav = this.root.find('nav');
				this.navButtons = this.nav.find('li');

				var liWidth = this.navButtons.eq(0).width();
				var liMargin = this.navButtons.eq(1).css('margin-left');
				liMargin = liMargin.substring(0, liMargin.length - 2);

				this.nav.find('ul').css({width: (liWidth * nbOfSlides) + (liMargin * (nbOfSlides - 1)) + "px"})

				this._setNavEvents();
			},

			//=====================================================================
			// _initSlides : Private Function
			//
			// @params : e
			//		If not defined, it means no click event was done to get here
			//		therefore, the direction should be 1, which is equal to right
			//		to left. Please note that not all of the animations will make
			//		use of the direction parameter. i.e. Fade in/out
			//
			//		If e is defined, a click was made. If it was on the nav, the
			//		direction property will be of 0. The nextIndex will then become
			//		the targeted nav button. Otherwise, the direction will be equal
			//		to what the arrow event set it to [see _setArrowEvents()]
			//=====================================================================
			_initSlides : function(e){
				this.currentSlide = this.slides.eq(this.currentIndex);
				this.nextIndex = this.currentIndex + this.direction;

				if(e){
					// If this direction == 0, it means that we clicked on the nav
					// buttons
					if(this.direction == 0) {
						this.nextIndex = this.navButtons.index($(e.target));

						if(this.nextIndex < this.currentIndex)
							this.direction = -1;
						else
							this.direction = 1;
					}
				}
				else {
					this.direction = 1;
				}

				if(this.nextIndex >= this.slides.length)
					this.nextIndex = 0;
				else if (this.nextIndex < 0)
					this.nextIndex = this.slides.length - 1;

				this.nextSlide = this.slides.eq(this.nextIndex);

				this.slides.hide();
				this.currentSlide.show();
			},

			//=====================================================================
			// _changeSlide : Private Function
			//
			// This is where the slides are changed. For the whole process, the 
			// arrows and the nav buttons will be disabled. They will be enabled
			// again after the animations are done. Depending on the transitionType
			// property of the slideshow, it will call the right animation function
			//=====================================================================
			_changeSlide : function(){
				var animation;

				if(this.params.transitionType == 'slide')
					animation = this._slide();
				else
					alert('Only the SLIDE effect is working right now.');

				animation.done(function(){
					if(this.params.hasNav){
						this._updateNav();
						this._setNavEvents();
					}

					if(this.params.hasArrows)
						this._setArrowEvents();
					
					if(this.params.timer)
						this._setTimer();
				}.bind(this))
			},

			//=====================================================================
			// _slide : Private Function
			//
			// @returns : $.Deferred();
			//
			// Function to be called when the transitionType property is set to
			// 'slide'.
			//=====================================================================
			_slide : function(){
				var firstSlide = $.Deferred();
				var secondSlide = $.Deferred();
				var animationDeferred = $.Deferred();

				this.nextSlide
					.css({left: (this.params.slideWidth * this.direction)})
					.show();
					
				TweenMax.to(this.currentSlide, this.params.transitionTime, {
					left: "+="+(this.params.slideWidth * (this.direction * -1)),
					
					onComplete:function(){
						firstSlide.resolve();

						this.currentSlide
							.hide()
							.css({left:0});
					}.bind(this)
				});

				TweenMax.to(this.nextSlide, this.params.transitionTime, {
					left: "+="+(this.params.slideWidth * (this.direction * -1)),
					
					onComplete:function(){
						secondSlide.resolve();
					}
				});

				$.when(firstSlide, secondSlide).then(function(){
					this.currentIndex = this.nextIndex;

					animationDeferred.resolve();
				}.bind(this));

				return animationDeferred;
			},

			//=====================================================================
			// _setTimer : Private Function
			//
			// Sets the timeout function to call _initSlides after each tick.
			//=====================================================================
			_setTimer : function(){
				this.timer = setTimeout(function(){
					this._initSlides();
					this._changeSlide();
				}.bind(this), this.params.timer);
			}
		}

		return Neutrino;
	}
);